File: ruby/scrabble-score/scrabble_score.rb

class Scrabble
  def self.score(word)
    Scrabble.new(word).score
  end

  def initialize(word)
    @word = word.to_s.gsub(/\W/, '').downcase
  end

  def score
    @word.chars.map(&method(:points)).reduce(0, &:+)
  end

  private

  def points(c)
    POINTS.find { |letters, p| letters.include? c }.last
  end

  POINTS = {
    'aeioulnrst' => 1,
    'dg' => 2,
    'bcmp' => 3,
    'fhvwy' => 4,
    'k' => 5,
    'jx' => 8,
    'qz' => 10
  }
end

File: clojure/scrabble-score/src/scrabble_score.clj
(ns scrabble-score)

(def raw-score {"AEIOULNRST" 1
                "DG"         2
                "BCMP"       3
                "FHVWY"      4
                "K"          5
                "JX"         8
                "QZ"         10})


(def score
  (into {} (for [[letters score] raw-score
                 letter letters]
            [letter score])))

(defn score-letter
  [letter]
  (score (clojure.string/upper-case letter)))

(defn score-word
  [word]
  (reduce + 0 (map score-letter word)))
